### PR TITLE
Fix auto-scroll should not pause on invisible activities

### LINKED ISSUE #3943: "Pause auto-scroll" doesn't play well with AC with Action.Submit/PostBack buttons
## Screenshots

recording (19)

## Version

`4.13.0`

## Describe the bug

With `styleOptions.autoScrollSnapOnActivity` set to `1`, after clicking on a submit button of Adaptive Cards, it sent a "post back" message. The transcript should scroll to the bottom.

Looking at the screenshot, after clicking on the bottom, the transcript should scroll to the bottom.

## Steps to reproduce

<details>
<summary>Expand to see test case</summary>

```js
const directLine = await testHelpers.createDirectLineWithTranscript(
  [
    createActivity(
      'Lorem culpa excepteur quis commodo qui mollit. Laboris laboris excepteur esse fugiat sunt id eu esse id fugiat sint. Veniam qui quis mollit ad magna elit ex laborum occaecat labore.'
    ),
    createActivity(
      'Aute deserunt id ullamco excepteur excepteur sint magna deserunt quis. Pariatur excepteur minim aliquip laboris quis eiusmod elit ut. Commodo dolor eu enim sint ex ex voluptate. Consequat duis laboris ad tempor aliqua aliquip et elit dolore do ea. Deserunt aute sit magna dolor sit elit dolor anim. Et reprehenderit adipisicing ea aute et exercitation est excepteur tempor cupidatat occaecat eiusmod eu.'
    ),
    createActivity(
      'Minim deserunt ut tempor minim eiusmod duis labore minim occaecat duis adipisicing proident cupidatat nisi. Aute aliquip qui culpa incididunt. Mollit consequat reprehenderit ex velit ipsum eu anim reprehenderit ad eiusmod veniam.'
    ),
    createActivity(
      'Eiusmod magna eiusmod proident dolor mollit mollit sit cupidatat consectetur est exercitation labore do quis. Esse incididunt duis consectetur consequat aliquip. Exercitation proident ut reprehenderit irure minim incididunt in sint anim.'
    ),
    createActivity(
      'Id cillum ut non quis commodo ex officia consequat duis mollit occaecat proident anim aliqua. Ex commodo anim adipisicing id fugiat adipisicing ut dolor nulla culpa aliquip adipisicing sit laboris. Esse non Lorem aute veniam quis esse quis ipsum reprehenderit deserunt ut do culpa ut. Laborum officia consectetur ad et non aute et nostrud reprehenderit deserunt elit sint.'
    ),
    {
      attachments: [
        {
          contentType: 'application/vnd.microsoft.card.adaptive',
          content: {
            $schema: 'http://adaptivecards.io/schemas/adaptive-card.json',
            type: 'AdaptiveCard',
            version: '1.3',
            body: [
              {
                type: 'TextBlock',
                text: 'Adaptive Card design session',
                size: 'Large',
                weight: 'Bolder',
                wrap: true
              },
              {
                type: 'TextBlock',
                text: ' Conf Room 112/3377 (10) ',
                isSubtle: true,
                wrap: true
              },
              {
                type: 'TextBlock',
                text: '12:30 - 01:30',
                isSubtle: true,
                spacing: 'None',
                wrap: true
              },
              {
                type: 'TextBlock',
                text: 'Snooze for',
                wrap: true
              },
              {
                type: 'Input.ChoiceSet',
                id: 'snooze',
                value: '5',
                choices: [
                  {
                    title: '5 minutes',
                    value: '5'
                  },
                  {
                    title: '15 minutes',
                    value: '15'
                  },
                  {
                    title: '30 minutes',
                    value: '30'
                  }
                ]
              }
            ],
            actions: [
              {
                type: 'Action.Submit',
                title: 'Snooze'
              },
              {
                type: 'Action.Submit',
                title: "I'll be late"
              }
            ]
          }
        }
      ],
      from: {
        id: 'bot',
        role: 'bot'
      },
      id: '1',
      timestamp: new Date().toISOString(),
      type: 'message'
    }
  ],
  {
    overridePostActivity: activity => {
      console.log(activity);

      const id = Math.random().toString(36).substr(2, 5);

      directLine.activityDeferredObservable.next({
        ...activity,
        id,
        timestamp: new Date().toISOString()
      });

      return Observable.from([id]);
    }
  }
);

WebChat.renderWebChat(
  {
    directLine,
    store: testHelpers.createStore(),
    styleOptions: {
      autoScrollSnapOnActivity: 1
    }
  },
  document.getElementById('webchat')
);

await pageConditions.uiConnected();
await pageConditions.numActivitiesShown(6);
await pageConditions.scrollToBottomCompleted();

// We need to send a message to mark all activities as read.
await host.click(document.getElementsByClassName('ac-pushButton')[0]);

directLine.activityDeferredObservable.next(
  createActivity(
    'Minim voluptate officia nisi excepteur pariatur tempor duis enim aute consectetur ex ipsum id. Exercitation voluptate proident sunt ad Lorem tempor ex dolore culpa aliquip ipsum eu. Sunt nisi esse esse consequat mollit exercitation incididunt incididunt deserunt culpa incididunt consectetur in. Magna proident aliquip sunt proident in velit cillum commodo minim proident cillum officia mollit consequat.'
  )
);

await pageConditions.scrollToBottomCompleted(); // Failed on this line
```
</details>

## Expected behavior

It should scroll to bottom after receiving a response from the bot, based on the `postBack` activity.

## Additional context

[Bug]

### PR DESCRIPTION
**Pull request by compulim**
> Fixes #3943.

### Fixed

- Fixes [#3943](https://github.com/microsoft/BotFramework-WebChat/issues/3943). Auto-scroll should skip invisible activities, such as post back or event activity, by [@compulim](https://github.com/compulim), in PR [#3945](https://github.com/microsoft/BotFramework-WebChat/pull/3945)

## Description

When we compute when to pause the scroll, we based on the `offsetTop` of "last acknowledged activity" (LAA). When the end-user send a message, the LAA will be that message.

If the LAA is an invisible activity, such as post back, message back, or event activity, we will not able to get the `offsetTop` because the LAA is not rendered.

This fix is to compute the `offsetTop` based on the "first visible activity after the LAA", instead of LAA itself.

## Specific Changes

- Modify the `scroller` callback to find out "first visible activity after the LAA" and use it for all computations
- Add 3 tests to make sure invisible activities from both side (bot/user) works with both "pause auto-scroll behavior" (num activities or num pages)

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] ~I have updated documentation~

## Review Checklist

> This section is for contributors to review your work.

-  [x] ~Accessibility reviewed (tab order, content readability, alt text, color contrast)~
-  [x] ~Browser and platform compatibilities reviewed~
-  [x] ~CSS styles reviewed (minimal rules, no `z-index`)~
-  [x] ~Documents reviewed (docs, samples, live demo)~
-  [x] ~Internationalization reviewed (strings, unit formatting)~
-  [x] ~`package.json` and `package-lock.json` reviewed~
-  [x] ~Security reviewed (no data URIs, check for nonce leak)~
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

File: __tests__/html/autoScroll.eventActivity.html [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            createActivity(
+              'Eiusmod magna eiusmod proident dolor mollit mollit sit cupidatat consectetur est exercitation labore do quis. Esse incididunt duis consectetur consequat aliquip. Exercitation proident ut reprehenderit irure minim incididunt in sint anim.'
+            ),
+            createActivity(
+              'Id cillum ut non quis commodo ex officia consequat duis mollit occaecat proident anim aliqua. Ex commodo anim adipisicing id fugiat adipisicing ut dolor nulla culpa aliquip adipisicing sit laboris. Esse non Lorem aute veniam quis esse quis ipsum reprehenderit deserunt ut do culpa ut. Laborum officia consectetur ad et non aute et nostrud reprehenderit deserunt elit sint.'
+            )
+          ],
+          {
+            overridePostActivity: activity => {
+              // console.log(activity);
```
Comment (cwhitten) [inline __tests__/html/autoScroll.eventActivity.html:43]: remove?

File: packages/component/src/BasicTranscript.js [hunk 2]
```diff
@@ -1038,41 +1041,58 @@ const SetScroller = ({ activityElementsRef, scrollerRef }) => {
...
-          if (firstUnacknowledgedActivityElement) {
-            values.push(
-              firstUnacknowledgedActivityElement.offsetTop -
-                scrollTop -
-                offsetHeight * (1 - patchedAutoScrollSnapOnPage) +
-                patchedAutoScrollSnapOnPageOffset
-            );
+          if (~firstUnacknowledgedActivityElementIndex) {
+            if (patchedAutoScrollSnapOnActivity) {
+              // Find the activity element which we should snap to.
```
Comment (cwhitten) [inline packages/component/src/BasicTranscript.js:1068]: Nit on use of "Find" in the comment. We're not "finding" in the known sense, but instead are just accessing the location where it sits, correct?

Comment (cwhitten) [inline packages/component/src/BasicTranscript.js:1068]: Nit on use of "Find" in the comment. We're not "finding" in the known sense, but instead are just accessing the location where it sits, correct?

**Review by cwhitten — COMMENTED** [review]
a couple comments

File: packages/component/src/BasicTranscript.js (hunk 2 above)
Comment (compulim) [reply packages/component/src/BasicTranscript.js:1068]: Right.

```
// Gets the activity element which we should snap to.
```

Comment (compulim) [reply packages/component/src/BasicTranscript.js:1068]: Right.

```js
// Gets the activity element which we should snap to.
```

**Comment by compulim** [comment]
Done. Ready for another round.

**Review by cwhitten — APPROVED** [review]

**Pull request merged**